### ISSUE TITLE
feat(examples): port alexnet_cifar10 to autograd path (sub-task 3 of #5454)

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#5559]
Sub-task **3 of 7** of #5454 — port every manual-gradient convnet example onto the autograd path (`Variable` / `GradientTape` / `optimizer.step(params, tape)`).

## Scope

Port `examples/alexnet_cifar10/` (**AlexNet-mini**) off the manual forward + `*_backward` + `model.update_parameters(lr, ...)` pattern onto the autograd substrate.

- **Approx LOC delta:** -300 / +120
- **Notes:** More layers, but same shape as the LeNet port.

## Prerequisite — satisfied

The blocker #5452 (`feat(autograd): complete autograd substrate for convnet training`) is **CLOSED**. The substrate this port needs is present in `src/projectodyssey/autograd/`:
`variable_conv2d`, `variable_maxpool2d`, `variable_linear`, `variable_flatten`, `variable_cross_entropy` (all in `variable.mojo`), and `tape.backward()` reverse-topological dispatch (`tape.mojo:296`). So this sub-task is **unblocked**.

## Port checklist (per #5454)

1. **Fork** `examples/alexnet_cifar10/` into an autograd variant first — do **not** delete the manual version until tests pass.
2. **Parity smoke run:** confirm loss/accuracy parity with the manual version on a subset (1 epoch), using a genuine in-container run (no fabricated metrics — committed logs are not evidence).
3. **Default optimizer → AdamW** (matches `examples/grok/lenet_emnist/`); keep SGD selectable.
4. Once **all 7 ports + the grok example** are autograd-based (tracked in #5454), a follow-up deletes the manual implementations and the `model.update_parameters(lr, ...)` pattern from the model classes.

## Definition of done

- [ ] Autograd variant of `examples/alexnet_cifar10/` trains via `GradientTape` + `optimizer.step`.
- [ ] Parity smoke run passes in-container (loss/accuracy within tolerance of the manual version).
- [ ] Default optimizer is AdamW; SGD still selectable.
- [ ] Example's tests run in CI (not merely compiled) — mirror the `example-backward-tests` gate pattern where applicable.
- [ ] Manual `*_backward`/`update_parameters` left intact until the tracker-level cleanup step.

## References

- Tracker: #5454 (suggested order 1→…→7; this is **3**).
- Sub-task 1: #5461 (`lenet_emnist` — closed, reference port).
- `src/projectodyssey/autograd/README.md`, `src/projectodyssey/autograd/optimizers.mojo` (`AdamW`).
- #5465 — a real autograd bug (conv2d backward kernel grads vanished in deep stacks) surfaced during the lenet port; watch for it in the deeper convnets.

**Comment by mvillmow** [comment]
# Implementation Plan — port `examples/alexnet_cifar10/` to the autograd path (sub-task 3 of #5454)

> Grounded in verified source. Every cited signature/layer-order was Read on disk; two load-bearing facts (no `variable_dropout` op; the exact conv/pool placement) were re-confirmed before posting. Uses the just-merged mnist port (PR #5585, GRADE A) as the direct template.

## Summary
Add `examples/alexnet_cifar10/run_train_autograd.mojo` — trains `AlexNet` on CIFAR-10 through the tape autograd substrate (`Variable`/`GradientTape`/`variable_*` + the `Optimizer` trait), mirroring `examples/mnist/train_autograd.mojo`. Default **AdamW**, `--optimizer sgd` selectable. `run_train.mojo` and `model.mojo` are NOT modified. Wire the new entry into the training-smoke CI matrix + `training-smoke-all`.

## Verified architecture (Read, not assumed)

**AlexNet has exactly 16 params, in this field order** (`model.mojo` struct + save/load agree):
`conv1_kernel, conv1_bias, conv2_kernel, conv2_bias, conv3_kernel, conv3_bias, conv4_kernel, conv4_bias, conv5_kernel, conv5_bias, fc1_weights, fc1_bias, fc2_weights, fc2_bias, fc3_weights, fc3_bias`.

**Exact forward order** (traced from `run_train.mojo::compute_gradients`; re-confirmed). Pools after conv1, conv2, conv5 **only** (conv3/conv4 have no pool):

| Manual op | Autograd equivalent |
|---|---|
| conv1 → relu → **pool1** | `variable_conv2d` → `variable_relu` → `variable_maxpool2d` |
| conv2 → relu → **pool2** | same |
| conv3 → relu | `variable_conv2d` → `variable_relu` |
| conv4 → relu | `variable_conv2d` → `variable_relu` |
| conv5 → relu → **pool3** | `variable_conv2d` → `variable_relu` → `variable_maxpool2d` |
| reshape → flatten | `variable_flatten` |
| fc1 → relu → **dropout** | `variable_linear` → `variable_relu` → **OMITTED** |
| fc2 → relu → **dropout** | `variable_linear` → `variable_relu` → **OMITTED** |
| fc3 → cross_entropy | `variable_linear` → `variable_cross_entropy` |

Import stride/padding + pool constants from `model` exactly as `run_train.mojo` (`CONV1_STRIDE`…`CONV5_PADDING`, `POOL1_KERNEL_SIZE`…`POOL3_PADDING`).

## Dropout gap (verified, must-address)
`grep` finds **no `variable_dropout` and no `OP_DROPOUT`** — the autograd substrate has no dropout op. The manual forward applies dropout twice (after fc1-relu and fc2-relu). **Resolution: omit both** in the autograd training forward (FC tail = fc1→relu→fc2→relu→fc3). This matches an inference-mode / non-regularized forward and mirrors the dropout-free mnist template (GRADE A). Document in the module docstring. Do NOT invent a `variable_dropout` call — it does not exist.

## AdamW-default trap (repeat of mnist review C1 — do NOT repeat)
The shared parser registers `optimizer` with default `"sgd"`, and `parse()` pre-populates it, so `get_string("optimizer","adamw")` returns `"sgd"` when unset (dead fallback) → default silently becomes SGD. **This was the CRITICAL that failed mnist's first review.** Fix identically:
```
var optimizer: String = "adamw"
if args.was_user_supplied("optimizer"):
    optimizer = args.get_string("optimizer", "adamw")
```
plus `use_adamw(o) -> Bool: return o != "sgd"`. `was_user_supplied` confirmed at `arg_parser.mojo:108`.

## New file: `examples/alexnet_cifar10/run_train_autograd.mojo`
Entry name matches convention (alexnet's smoke entry is `run_train.mojo` → autograd sibling `run_train_autograd.mojo`). Fork `run_train.mojo`, keep its CIFAR-10 load + `--smoke` synthetic branch + `--max-batches` + save-guard; replace manual `compute_gradients` with the autograd `train_batch`/`train_epoch` from mnist.

1. **Imports:** `AlexNet` + 16 constants from `model`; `Variable, GradientTape, AdamW, SGD, variable_conv2d, variable_relu, variable_maxpool2d, variable_flatten, variable_linear, variable_cross_entropy` from `projectodyssey.autograd`; `Optimizer` from `optimizer_base`; keep `CIFAR10Dataset`/`one_hot_encode`/`AnyTensor`/`zeros`/`create_training_parser`. Drop manual-only imports (`*_backward`, `dropout`, `cross_entropy`, momentum/scheduler helpers).
2. **`TrainConfig` + `parse_args`** (mnist shape): fields `epochs, batch_size, learning_rate, weight_decay, optimizer, data_dir, weights_dir, max_batches, smoke`; `weights-dir` default `"alexnet_weights_autograd"`, `data-dir` `"datasets/cifar10"`; defaults epochs 100 / batch 128 / lr 0.01 / wd 0.01. Resolve optimizer via the `was_user_supplied` gate. Manual `momentum` arg dropped (note in docstring).
3. **`use_adamw`** — copy mnist verbatim.
4. **`train_batch[O: Optimizer](mut model: AlexNet, input, labels, mut optimizer: O, mut tape) -> Float32`:** wrap 16 params `requires_grad=True` in field order, inputs `False`; record the forward chain (dropout omitted); `loss.backward(tape)`; `loss_value = loss.data._data.bitcast[Float32]()[0]` **before** moving params; append 16 params `^` in field order; `optimizer.step`; write back `model.<field> = parameters[i].data` **same order** (0..15); `optimizer.zero_grad(tape)`.
5. **`train_epoch[O: Optimizer]`:** fresh `tape` per batch. **Batch extraction: reuse `run_train.mojo`'s `.slice(start,end,axis=0)`** (dtype-safe, the proven alexnet path — no need for mnist's `_get_float64` copy loop, which existed only to avoid a raw 1-byte pointer bug that `.slice` doesn't have). `one_hot_encode(..., num_classes=model.num_classes)`; `max_batches` cap; `% 100 == 0 or num_batches <= 10` logging.
6. **`main()`:** banner incl. `"Optimizer:", "AdamW" if use_adamw(name) else "SGD"`; `AlexNet(num_classes=..., dropout_rate=0.5)` (dropout_rate stored, unused); keep the `if smoke:` synthetic branch **verbatim** (3×32×32 f32, uint8 `[N]`, `wanted_batches = max_batches if >0 else 3`) + `else` `CIFAR10Dataset`; construct optimizer **once per branch** (AdamW/SGD) so AdamW moments persist; drop `step_lr` (note omission); optional per-epoch eval only if it compiles clean, else skip (smoke asserts no accuracy); `if not smoke: model.save_weights(...)`.

## CI wiring
- `.github/workflows/comprehensive-tests.yml` training-smoke matrix: add `- { example: alexnet_cifar10,   entry: run_train_autograd.mojo }`.
- `justfile` `training-smoke-all` ENTRIES: add `"alexnet_cifar10 run_train_autograd.mojo"`.

## Verification (genuine in-container only)
**Anti-fabrication:** committed logs are never evidence; a truthful failure is acceptable, invented success is not.
1. Compiles clean via the smoke recipe's Werror+AVX path.
2. `just training-smoke-one alexnet_cifar10 run_train_autograd.mojo` → exit 0, ≥2 finite (non-nan/inf) losses.
3. Default (no `--optimizer`) banner → `AdamW`.
4. `--optimizer sgd` banner → `SGD`.
5. Real-data CIFAR-10 parity: check `ls datasets/cifar10/`; if absent, disclose parity **PENDING** (do NOT fabricate); if present, bounded `--max-batches` run vs manual `run_train.mojo`, finite decreasing loss.

## Risks
- **#5465 (5 stacked convs = heaviest conv-backward exposure yet):** CLOSED/fixed, but this is the first autograd path recording 5 `variable_conv2d` + 3 maxpools — validate **empirically** (check 2), don't assume.
- **AdamW-default trap:** the exact mnist C1; enforced by the `was_user_supplied` gate + checks 3–4.
- **Dropout gap:** documented omission (no regularization vs manual) — acceptable for a mechanism port.
- **16-param write-back order:** one transposition silently corrupts a layer; use one field-name list for both append + write-back, cross-check `model.mojo` save order.
- **Memory/runtime:** 5 convs + fc 4096×4096 on the tape is large; keep smoke small (default `wanted_batches=3`); watch actual wall-clock in check 2 vs the training-smoke budget.

## Unverified API Assumptions
| Assumption | Verify at |
|---|---|
| `just training-smoke-one` applies Werror + runs the entry | `justfile` `_training-smoke-one-inner` (~:1356) + check 2 |
| optional per-epoch `evaluate`/`model.predict` compiles in the autograd entry | compile check 1; omit if it errors |
| `DatasetInfo("cifar10").num_classes() == 10` | `data/constants.mojo`, or print `model.num_classes` (check 3) |
| `.slice` sliced batch feeds `variable_conv2d` correctly | finite loss in check 2 |
| CIFAR-10 dataset present for parity | `ls datasets/cifar10/` before check 5 |

## DoD → steps
- [ ] `run_train_autograd.mojo` created; `run_train.mojo`/`model.mojo` unchanged (`git diff --stat` = new file + 2 CI lines).
- [ ] Forward records the chain (dropout omitted); 16 params wrapped; write-back order == append order == `model.mojo` field order. → check 1.
- [ ] AdamW default via `was_user_supplied`; no dead fallback. → checks 3–4.
- [ ] Optimizer once per branch; `.slice` batching; `if not smoke:` save guard; dropout omission documented. → review + check 2.
- [ ] CI matrix + `training-smoke-all` updated. → grep both.
- [ ] `just training-smoke-one alexnet_cifar10 run_train_autograd.mojo` exit 0, ≥2 finite losses. → check 2.
- [ ] Banner AdamW default / SGD selectable. → checks 3–4.
- [ ] Real-data parity OR explicit PENDING disclosure. → check 5.

### Critical files
- `examples/alexnet_cifar10/run_train.mojo` (fork source), `examples/alexnet_cifar10/model.mojo` (16-param order + constants)
- `examples/mnist/train_autograd.mojo` (template)
- `src/projectodyssey/autograd/variable.mojo` (variable_* ops; NO variable_dropout)
- `.github/workflows/comprehensive-tests.yml` + `justfile` (CI wiring)
